cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -39,12 +43,336 @@
 namespace arrow {
 
 using internal::checked_cast;
+using internal::PrimitiveScalarBase;
 
 namespace compute {
 
 namespace {
 
-struct GrouperImpl : Grouper {
+constexpr uint32_t kNoGroupId = std::numeric_limits<uint32_t>::max();
+
+using group_id_t = std::remove_const<decltype(kNoGroupId)>::type;
+using GroupIdType = CTypeTraits<group_id_t>::ArrowType;
+auto g_group_id_type = std::make_shared<GroupIdType>();
+
+inline const uint8_t* GetValuesAsBytes(const ArraySpan& data, int64_t offset = 0) {
+  DCHECK_GT(data.type->byte_width(), 0);
+  int64_t absolute_byte_offset = (data.offset + offset) * data.type->byte_width();
+  return data.GetValues<uint8_t>(1, absolute_byte_offset);
+}
+
+template <typename Value>
+Status CheckForGetNextSegment(const std::vector<Value>& values, int64_t length,
+                              int64_t offset, const std::vector<TypeHolder>& key_types) {
+  if (offset < 0 || offset > length) {
```

Comment:
I think the entire function can be probably DCHECK - these checks are more like "invariants" to me. @westonpace WDYT?

Reply:
Changing this to a DCHECK breaks the error-checks in `TEST(RowSegmenter, Basics)`, so I won't make a change here for now. If you still want to change, you should say whether to defer on this, or to remove the error-checks, or to make internal APIs for the error-checks, or something else.

Reply:
Which error-check is this breaking? If this is a DCHECK, then it is invariant that shouldn't happen and I don't think we need to have specific test case for it.

Reply:
I think this should still be a DCheck here because I see the class here being "internal util that used for segment aggregation" and not a "library for users to use". But not a big issue so I wouldn't hold on to it.

I will create a follow up issue to track this

Reply:
Tracked in https://github.com/apache/arrow/issues/34529